pyro/autohelp/ast_visitor.py

```diff
@@ -0,0 +1,246 @@
+            return False
+        return res
+
+    def on_leave(self, original_node, updated_node):
+        if changes := self.updates.get(original_node):
+            return changes(updated_node)
+        return updated_node
+
+
+class _FindCommandDecorator(libcst.CSTVisitor):
+    def __init__(self):
+        super().__init__()
+        self.is_command_dec = False
+
+    def visit_Decorator(self, node):
+        if not isinstance(node.decorator, libcst.Call):
+            return True
+        func = node.decorator.func
+        # find decorators that are like @thing.slash_command
+        # but ensure to ignore @commands.command` since that should have self.`
+        if func.attr.value.endswith("command") and func.value not in (
+            "commands",
+            "nextcord",
+        ):
+            self.is_command_dec = True
+        return True
+
+
+class EventListenerVisitor(BaseHelpTransformer):
+    def visit_ClassDef(self, node: libcst.ClassDef):
+        # don't check classdefs for decorators
+        return False
+
+    def visit_Decorator(self, node: libcst.Decorator) -> None:
+        if (
+            isinstance(node.decorator, libcst.Call)
+            and node.decorator.func.attr.value == "event"
+        ):
+            self.found_errors.append(Actions.DECORATOR_EVENT_CALLED)
+            # there's two solutions here. This needs to either be a listener or an event with no call.
+            # this will be left up to implement later, for now, we convert it to just an uncalled event
+            self.updates[node] = lambda node: node.with_changes(
+                decorator=node.decorator.func
+            )
+        elif (
+            isinstance(node.decorator, libcst.Attribute)
+            and node.decorator.attr.value == "listen"
+        ):
+            self.found_errors.append(Actions.DECORATOR_LISTEN_NOT_CALLED)
+            self.updates[node] = lambda node: node.with_changes(
+                decorator=libcst.Call(node.decorator)
+            )
+
+    def visit_FunctionDef_params(self, node: libcst.FunctionDef) -> Optional[bool]:
+        if not node.params or not node.decorators:
+            return False
+        first_param: libcst.Param = node.params.params[0]
+        if not first_param.name.value == "self":
+            return False
+        # its self and a top level method, so check for a bot decorator
+        # needs a decorator with `command` in the last section and to be called
+        visitor = _FindCommandDecorator()
+        node.visit(visitor)
+        if visitor.is_command_dec:
+            self.found_errors.append(Actions.USING_SELF_ON_BOT_COMMAND)
+
+            def update(node: libcst.FunctionDef):
+                params = node.params.with_changes(params=node.params.params[1:])
+                return node.with_changes(params=params)
+
+            self.updates[node] = update
+
+
+class ClientIsNotBot(BaseHelpTransformer):
+
+    BOT_CLASSES = ("Bot", "InteractionBot")
```

Comment:
How hard would it be to support subclasses do you think? Does Libcst expose enough information to check?

If its not easy peasy, don't worry about it.

Reply:
Not super hard, but the code would have to be present that is subclassing the Bot class. I can't deduce that by name alone.

Reply:
Fair, dont worry then